feat: 메인 페이지 UI 구현

devjaeyoon opened the pull request:
### ✨ 이슈 번호

- #16 

### 📌 설명

메인 페이지의 UI를 구현하여 작성한 Pull request입니다.

### 📃 작업 사항

- [x]  **Favorite Channel**  영역
    - [x]  별(★) 아이콘이 색칠된 채로 표시되어야 한다.
    - [x]  가로 스크롤 카드 형태 UI이다.
    - [x]  채널 썸네일과 이름을 포함한다.
    - [ ] ~별 (☆) 아이콘 클릭 시 Channel로 이동한다.~
    - [x]  순서 조정이 가능하다 (`::` 아이콘 롱프레스 및 드래그 가능)
- [x]  **All** 채널 목록
    - [x]  직사각형 카드 형태 UI이다.
    - [ ]  ~별 (☆) 아이콘 클릭 시 Favorite Channel로 이동한다.~
    - [x]  썸네일과 이름을 포함한다.
- [x] 즐겨찾기 채널 목록과 일반 채널 목록을 분리된 UI로 구현
- [x] @dnd-kit을 활용해 즐겨찾기 채널의 정렬 기능 추가
- [x] 컴포넌트 역할에 따라 UI 및 로직 구조를 명확히 분리
- [x] 커스텀 훅으로 상태 분류 기능 기초 구현  

### 💡 참고 사항

[BlankStar, FilledStar, DragDrop svg 파일의 파일명을 파스칼 케이스에서 케밥 케이스로 변경했습니다.](https://github.com/Radio-Premium/RadioPremium-FE/pull/11#discussion_r2111863583)

Favorite Channel에서 별 아이콘 클릭 시 Channel로 이동하는 기능, Channel에서 별 아이콘 클릭 시 Favorite Channel로 이동하는 기능은 백엔드와의 API 연동 작업에서 진행해야 하므로 작업하지 않았습니다.

### 📁 변경 파일 및 구성 설명

1. useCategorizeChannel.jsx
 • 채널 데이터를 즐겨찾기 여부 기준으로 분류하여 반환하는 커스텀 훅입니다.
 • 현재는 초기 구조만 구현되어 있으며, 추후 API 연동 시 즐겨찾기 채널과 일반 채널을 자동 분류할 수 있도록 설계되어 있습니다.
 • 반환값: [일반 채널 목록, 즐겨찾기 채널 목록]

2. Home.jsx
 • 메인 페이지 컴포넌트입니다.
 • 상단에 여백 및 TabBar가 위치하고, 그 아래에 두 개의 섹션이 배치됩니다:
 • 즐겨찾기 채널 목록 섹션 (FavoriteChannelList)
 • 일반 채널 목록 섹션 (ChannelListItem)
 • useCategorizeChannels 훅을 통해 데이터를 받아와 두 섹션에 분리 렌더링합니다.
 • UI는 기본 패딩, 스크롤, 타이포그래피가 적용되어 전체적인 페이지 구조가 구성되어 있습니다.

3. FavoriteChannelList.jsx
 • @dnd-kit의 DndContext와 SortableContext를 이용해 드래그 앤 드롭으로 채널 순서 변경이 가능한 컴포넌트입니다.
 • favoriteChannelList를 arrayMove로 재정렬하며 상태를 관리합니다.
 • restrictToVerticalAxis와 커스텀 Y축 범위 제한(restrictToYRange)을 통해 수직 이동만 가능하도록 제어합니다.
 • 각 아이템은 FavoriteChannelListItem 컴포넌트로 구성됩니다.

4. FavoriteChannelListItem.jsx
 • useSortable 훅을 사용하여 드래그 가능한 채널 아이템을 구현한 컴포넌트입니다.
 • 채널 이름, 썸네일, 즐겨찾기 아이콘, 드래그 핸들을 포함합니다.
 • setActivatorNodeRef를 드래그 아이콘에 연결해 해당 아이콘으로만 드래그되도록 설정되어 있습니다.
 • 시각적으로 구분된 아이템 UI가 적용되어 있으며, 향후 클릭 이벤트 등을 연동하는 구조입니다.

5. ChannelListItem.jsx
 • 일반 채널 목록에 사용되는 리스트 아이템 컴포넌트입니다.
 • 즐겨찾기 여부에 따른 분기 로직을 제거하여 일반 채널 전용 컴포넌트로 단순화하였습니다.
 • 채널 썸네일, 이름, 즐겨찾기 버튼(BlankStarIcon)으로 구성되어 있습니다.
 • onToggleFavorite prop을 통해 외부에서 즐겨찾기 상태 변경을 제어할 수 있도록 설계되어 있습니다.

### 🔁 예상 흐름 요약
	
1. useCategorizeChannels 훅을 통해 채널 데이터를 즐겨찾기 여부에 따라 분리
2. Home.jsx에서 해당 데이터를 받아 각각 FavoriteChannelList, ChannelListItem으로 전달
3. FavoriteChannelList는 드래그 앤 드롭 기능이 포함된 리스트로 정렬 가능
4. ChannelListItem은 즐겨찾기 등록을 위한 버튼만 포함한 일반 목록 UI로 구성
5. 각 채널 아이템은 독립적인 컴포넌트로 관리되어 유지보수 및 확장이 용이한 구조

### 💭 리뷰 사항 반영

- [x] 파일 끝 개행 추가
- [x] props를 useState의 초기값으로 복사하면 props가 변경되어도 갱신되지 않는 문제 useEffect의 deps로 이용하여 해결
- [x] 공통된 레이아웃 ChannelSection 컴포넌트로 분리 및 리팩토링
- [x] 채널명이 길 경우 레이아웃이 깨질 수 있는 문제 해결
- [x] DragDrop 버튼 cursor-grab/grabbing 추가  

### ✅ Pull Request 체크 사항

- [x] 가장 최신 브랜치를 pull 하였습니다.
- [x] base 브랜치명을 확인하였습니다.
- [x] 코드 컨벤션을 모두 확인하였습니다.
- [x] 브랜치명을 확인하였습니다.

devjaeyoon committed:
rename: 파스칼 케이스 아이콘명 케밥 케이스로 변경

devjaeyoon committed:
style: 전역 스크롤바 제거 및 터치 스크롤 비활성화

Tailwind import에 세미콜론 추가하고, html과 body에 overflow: hidden 및
touch-action: none 설정하여 스크롤 및 드래그 영역 제어.
.scrollbar-hide 클래스를 추가해 브라우저별 스크롤바 숨김 처리.

devjaeyoon committed:
style: MainLayout에 wrapper div 추가 및 전체 레이아웃 크기 설정

- main 바깥에 div를 추가하여 전체 높이/너비를 h-dvh, w-screen으로 설정
- 전역 레이아웃 크기 일관성 확보 및 overflow 제어를 위한 준비 작업

devjaeyoon committed:
chore: @dnd-kit 관련 의존성 패키지 설치 및 설정

- @dnd-kit/core, @dnd-kit/sortable, @dnd-kit/modifiers 설치
- 드래그 앤 드롭 기능 구현을 위한 기본 환경 구성

devjaeyoon committed:
style: 아이콘 파일명 변경 및 ChannelListItem 레이아웃 수정

- kebab-case로 변경된 SVG 파일명 컴포넌트 내 반영
- flex 레이아웃 개선

devjaeyoon committed:
feat: 즐겨찾기 기준 채널 분류 훅 기본 구조 구현

향후 API 연동을 통해 즐겨찾기 채널과 일반 채널을 분리할 수 있도록
기초 상태와 초기화 로직을 포함한 커스텀 훅을 미리 구현함

devjaeyoon committed:
feat: 즐겨찾기 채널 리스트 드래그 앤 드롭 정렬 기능 구현

@dnd-kit을 활용하여 즐겨찾기 채널 목록의 우선순위를 변경할 수 있도록
드래그 앤 드롭 기능이 적용된 FavoriteChannelList 컴포넌트 구현

devjaeyoon committed:
feat: 즐겨찾기 채널 아이템 컴포넌트 구현

@dnd-kit의 useSortable 훅을 활용해 드래그 가능한 리스트 아이템 구성
아이콘 및 스타일을 적용하여 UI 구성 완료

devjaeyoon committed:
refactor: ChannelListItem에서 즐겨찾기 여부 분기 제거

즐겨찾기 채널과 일반 채널 목록 컴포넌트를 분리함에 따라
ChannelListItem은 일반 채널 전용 컴포넌트로 단순화

devjaeyoon committed:
feat: 메인 페이지 UI 구현 및 채널 목록 레이아웃 분리

즐겨찾기 채널과 일반 채널을 구분하여 각각 섹션 구성
TabBar 및 기본 패딩 포함한 전체 페이지 레이아웃 구현

d0-neii commented on src/index.css:
```diff
@@ -1 +1,17 @@
+}
+.scrollbar-hide::-webkit-scrollbar {
+  display: none;
+}
```
개행 추가 부탁드려요!

devjaeyoon replied:
추가했습니다. 말씀해주셔서 감사합니다~

devjaeyoon committed:
chore: 파일 끝 개행 추가

faber-h commented on src/components/FavoriteChannelList.jsx:
```diff
@@ -0,0 +1,61 @@
+
+import FavoriteChannelListItem from "@/components/FavoriteChannelListItem";
+
+const FavoriteChannelList = ({ channelList }) => {
```
props를 useState의 초기값으로 복사하면 props가 변경되어도 갱신되지 않는 문제가 생길 수 있어요~
수정 부탁드려요~

devjaeyoon replied:
Home 컴포넌트에서 업데이트 되더라도 FavoriteChannelList 내부의 favoriteChannelList는 처음 받은 값만 가지고 있어서 상위 상태와 하위 상태의 불일치가 일어날 수 있겠네요! `useEffect`로 `prop` 변화를 감지해서 동기화하는 방향으로 진행하겠습니다. 좋은 리뷰 감사드립니다!

faber-h commented on src/pages/Home.jsx:
```diff
@@ -1,5 +1,37 @@
+      <div className="h-12 w-full" />
+      <TabBar />
+      <div className="px-4 pt-8">
+        <section>
```
section, h2, ul 구조가 Favorite Channel, Channel에서 반복되고 있어요~
공통된 레이아웃은 컴포넌트로 추출하면 가독성과 유지보수성이 좋아질 것 같아요~

devjaeyoon replied:
생각하지 못한 점 짚어주셔서 감사합니다! 반영하겠습니다~

d0-neii commented on src/components/FavoriteChannelListItem.jsx:
```diff
@@ -0,0 +1,43 @@
+      className="my-2 flex h-16 w-full items-center rounded-md bg-gray-100 p-2 select-none"
+    >
+      <div className="h-12 w-12 flex-none rounded-lg bg-gray-200" />
+      <p className="ml-3 w-3/5 flex-1 text-sm font-bold">{name}</p>
```
채널 명이 예상보다 길어질 것을 대비해 `truncate`같은 속성을 추가하면 더 좋을 것 같아요!

devjaeyoon replied:
짚어주셔서 감사합니다! 보완했습니다!

devjaeyoon committed:
refactor: Home 페이지 채널 영역 공통 레이아웃 컴포넌트로 분리

d0-neii commented on src/components/FavoriteChannelListItem.jsx:
```diff
@@ -0,0 +1,43 @@
+        <button>
+          <FilledStarIcon className="mr-1 mb-0.5 ml-2" />
+        </button>
+        <button ref={setActivatorNodeRef} className="touch-none" {...listeners}>
```
지금은 커서 피드백이 없어서 웹에선 드래그가 가능하다는 걸 알아차리기 어려울 수 있을 것 같아요
`cursor-grab / grabbing` 같은 속성을 주면  UX적으로 더 직관적일 것 같습니다~

devjaeyoon replied:
가끔씩 마우스 모양이 바뀐 걸 본 적이 있었는데 이 속성이 `cursor-grab/ grabbing`인줄 몰랐네요! 알려주셔서 감사합니다! 반영했습니다~

devjaeyoon committed:
fix: FavoriteChannelList props 변경 시 상태 동기화 이슈 수정

props.channelList를 초기값으로만 사용하는 문제로 인해 상위 상태 변경이 반영되지 않던 문제 해결
useEffect로 props 변경을 감지하여 state를 동기화하도록 수정

devjaeyoon committed:
style: 드래그 중 아이콘 커서를 grabbing으로 변경

isDragging 상태를 활용하여 드래그 중일 때 커서를 cursor-grabbing으로,
그 외에는 cursor-grab으로 동적으로 지정

devjaeyoon committed:
style: 채널명이 길 경우를 대비해 텍스트 잘림 처리 추가

truncate, overflow-hidden, whitespace-nowrap 클래스로 UI 깨짐 방지

Eun0713 reviewed (approved):
LGTM!

faber-h reviewed (approved):
LGTM!
